### PR TITLE
Entity(Uri) Counts Logic

### PR DESCRIPTION
**Pull request by nmadhire**
Pull Request for reviewing Entity counts using Apache Spark and Scala.

### CODE REVIEW

File: wikipedia-stats-extractor.iml [hunk 1]
```diff
@@ -0,0 +1,183 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jodaiber) [inline wikipedia-stats-extractor.iml]: Please remove this file and add it to your .gitignore, this is IDE-only.

File: src/main/java/org/dbpedia/wiki/format/XmlInputFormat.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import org.apache.hadoop.mapreduce.lib.input.FileSplit;
+import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
+import org.dbpedia.spotlight.parser.JsonWikiParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+
```
Comment (dav009) [inline src/main/java/org/dbpedia/wiki/format/XmlInputFormat.java]: I guess this class was part of the experimentation but won't really be part of the code base

File: src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+package org.dbpedia.spotlight
```
Comment (dav009) [inline src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala]: Probably a better package name: `org.dbpedia.spotlight.XX` related to wikiparsing, stats extractions

File: src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+import org.apache.hadoop.io.{LongWritable, Text}
+import org.apache.spark.SparkContext
+import org.apache.spark.sql
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.JavaConversions._
+
+
+class WikipediaParser(inputWikiDump:String,lang:String)(implicit val sc: SparkContext){
```
Comment (dav009) [inline src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala]: can we rename this to `JsonpediaParser` ?

File: src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+    //Method to JsonParsing
+    val dfSurfaceForms = parseJson(sqlContext,dfWikiRDD)
+  }
+
+
+  /*
+    Method to parse the XML dump into JSON
+ */
+  private def readFile(path: String, sc: SparkContext): RDD[String] = {
```
Comment (dav009) [inline src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala]: I guess this is part of the first code expimernt. And can presumably be removed

File: src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala (hunk 4 above)
Comment (dav009) [inline src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala]: can we rename this to `JsonpediaParser` and make it such that it only contains the code that knows how to parse the jsonpedia i.e: reading the dataframes etc..

Comment (dav009) [inline src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala]: can we add a trait  that is called `wikipediaParser` and then `JsonpediaParser` implements `parse`. 
Dont need to make it too generic in this PR, but it eventually return a more abstract RDD back

File: src/main/java/org/dbpedia/util/DBpediaUriEncode.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+//import org.dbpedia.spotlight.util.WikiUtil;
+
+import org.dbpedia.extraction.util.WikiUtil;
+
+import java.io.Serializable;
+import java.util.regex.Pattern;
+/*
+Class used to Encode wiki Uri
+ */
+public class DBpediaUriEncode {
```
Comment (dav009) [inline src/main/java/org/dbpedia/util/DBpediaUriEncode.java]: does this class come from the spotlight project?

File: src/main/scala/org/dbpedia/spotlight/util/WikiUtil.scala [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    return encoded
+    //StringUtils.escape(encoded, replacements)
+  }
+
+
+  /**
+   * name encoded MediaWiki page name, e.g. '%C3%89mile_Zola'.
+   * Must not include the namespace (e.g. 'Template:').
+   */
+  /*def wikiDecode(name : String) : String =
```
Comment (dav009) [inline src/main/scala/org/dbpedia/spotlight/util/WikiUtil.scala]: is this really not needed?

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,7 @@
+.idea/
+target/
+.classpath
+.project
+.settings/
+.DS_STORE
+javadoc
```
Comment (dav009) [inline .gitignore]: weird EOF

File: pom.xml [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.specs</groupId>
+      <artifactId>specs</artifactId>
+      <version>1.2.5</version>
+      <scope>test</scope>
+    </dependency>
+      <dependency>
+          <groupId>org.apache.spark</groupId>
```
Comment (dav009) [inline pom.xml:60]: not sure how you are running this at the moment (via IDE?) 
but  you should add the tag provided to the spark dependencies

File: pom.xml [hunk 10]
```diff
@@ -0,0 +1,162 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Comment (dav009) [inline pom.xml:1]: Not for you to worry/address this at the moment:
But this pom file could be replaced by an sbt file with 3, 4 lines.

File: src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala (hunk 3 above)
Comment (jodaiber) [inline src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala]: How about

org.DBpediaSpotlight.spotlight.wikistats

?

File: src/main/java/org/dbpedia/wiki/format/XmlInputFormat.java (hunk 2 above)
Comment (nmadhire) [inline src/main/java/org/dbpedia/wiki/format/XmlInputFormat.java]: We need this. This is the class which reads each article XML from the raw wiki dump and creates article JSON as an element of RDD.

Comment (dav009) [inline src/main/java/org/dbpedia/wiki/format/XmlInputFormat.java]: oh ok, but Ideally this piece of code will be gone eventually when you plug in the code that handles jsonpedia?

Comment (nmadhire) [inline src/main/java/org/dbpedia/wiki/format/XmlInputFormat.java]: Already doing that in getCurrentValue() para. Calling JsonPedia logic..for every parsed XML coming out from raw wiki.

Comment (dav009) [inline src/main/java/org/dbpedia/wiki/format/XmlInputFormat.java]: :+1:

File: src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala (hunk 3 above)
Comment (nmadhire) [inline src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala]: org.DBpediaSpotlight.spotlight.wikistats 
or org.dbpedia.spotlight.wikistats

Do we need two spotlight in package names?

File: src/main/scala/org/dbpedia/spotlight/util/WikiUtil.scala (hunk 7 above)
Comment (nmadhire) [inline src/main/scala/org/dbpedia/spotlight/util/WikiUtil.scala]: Yes. This is not needed. I was able to call wikiutils from dbpedia-extraction repo.
I will remove this.

File: pom.xml (hunk 9 above)
Comment (nmadhire) [inline pom.xml:60]: Yes using IDE. I have spark installed on my laptop. It is working fairly well with small datasets.

File: src/main/java/org/dbpedia/util/DBpediaUriEncode.java (hunk 6 above)
Comment (nmadhire) [inline src/main/java/org/dbpedia/util/DBpediaUriEncode.java]: This one was in pignlproc for computing the dbpedia uri.. So I just copied this from there.

File: src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala (hunk 3 above)
Comment (jodaiber) [inline src/main/scala/org/dbpedia/spotlight/WikipediaParser.scala]: Ah, sorry, was using my phone. Only one spotl.

**Comment by nmadhire** [comment]
Organized the code little bit to make it more general. Let me know if you see something more can be improved.

File: src/main/scala/org/dbpedia/spotlight/wikistats/ComputeStats.scala [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    val language = lang
+
+    //Parse the individual WikiIds and create URI Counts
+    val dfSurfaceForms = dfWikiRDD.select("links.id")//.where("")
+      .rdd
+      .map(artRow => artRow.getList[String](0))
+      .flatMap(articleIds => articleIds.map(id=>id))
+      .mapPartitions{ wikiIds =>
+      val dbpediaEncode = new DBpediaUriEncode(language)
+      wikiIds.map(wikiId => (dbpediaEncode.uriEncode(wikiId),1))}
```
Comment (dav009) [inline src/main/scala/org/dbpedia/spotlight/wikistats/ComputeStats.scala]: indentation might help here

File: pom.xml (hunk 9 above)
Comment (dav009) [inline pom.xml:60]: be aware that if the generated jar is ever to be submitted to spark the `provided` tag will be needed

**Comment by dav009** [comment]
It looks much better after your changes.
Im willing to let this PR merged.

It would be good as part of the next batch of  work  to abstract how things are parsed from `ComputeStats`.

In this case `countURI` has to know about `link.ids` which is inherently related to the way jsonpedia is structured.

It would be good if the parser exposes certain interfaces, such that all of this internals are not required to be known by `ComputeStats`.

 for example counting uris could look like similar to:

``` scala
def countURI(){
    parser.getURIS().map(..function with count magic..)
}
```

**Comment by nmadhire** [comment]
This is good to go now. I will change the ComputeStats logic in the next PR tomorrow.

**Pull request merged**